tests/15_Versioning/VersionManagerTest.php

```diff
@@ -275,6 +282,113 @@ public function testRestorePendingChanges()
+        $this->vm->checkout($nodePath);
+
+        $node = $this->session->getNode($nodePath);
+        $node->setProperty('bar', 'value');
+
+        $this->session->save();
+
+        $this->assertTrue($node->hasProperty('bar'));
+
+        $this->vm->restore(true, $version);
+        $node = $this->session->getNode($nodePath);
+
+        $this->assertFalse($node->hasProperty('bar'));
+    }
+
+    public function testRestoreIsCheckedIn()
+    {
+        $nodePath = '/tests_version_base/versioned';
+        $version = $this->vm->checkpoint($nodePath);
+
+        $this->vm->checkout($nodePath);
+        $this->assertTrue($this->vm->isCheckedOut($nodePath));
+
+        $this->vm->restore(true, $version);
+        $this->assertFalse($this->vm->isCheckedOut($nodePath));
+    }
+
+    public function testRestoreBaseProperties()
+    {
+        // TODO also check for primary node type once it can be changed
+
+        $nodePath = '/tests_version_base/versioned';
+        $version = $this->vm->checkpoint($nodePath);
+        $this->vm->checkout($nodePath);
+
+        $node = $this->session->getNode($nodePath);
+        $node->addMixin('mix:created');
+
+        $this->session->save();
+
+        $node = $this->session->getNode($nodePath);
+        $this->assertContains('mix:created', $node->getPropertyValue('jcr:mixinTypes'));
+
+        $this->vm->restore(true, $version);
+
+        $node = $this->session->getNode($nodePath);
+        $this->assertEquals(array('mix:versionable'), $node->getPropertyValue('jcr:mixinTypes'));
+    }
+
+    public function testRestoreWithChildren()
+    {
+        $nodePath = '/tests_version_base/versioned';
+        $this->vm->checkout($nodePath);
+        $node = $this->session->getNode($nodePath);
+        $childNode1 = $node->addNode('childNode1');
+        $childNode1->setProperty('foo', 'child1');
+        $childNode2 = $node->addNode('childNode2');
+        $childNode2->setProperty('foo', 'child2');
+
+        $this->session->save();
+        $version = $this->vm->checkin($nodePath);
+
+        $this->assertCount(2, $node->getNodes());
+        $this->assertEquals('child1', $node->getNode('childNode1')->getPropertyValue('foo'));
+        $this->assertEquals('child2', $node->getNode('childNode2')->getPropertyValue('foo'));
+
+        $this->vm->checkout($nodePath);
+
+        $childNode1->remove();
+        $childNode2->setProperty('foo', 'child1');
+
+        $this->session->save();
+
+        $this->assertCount(1, $node->getNodes());
+        $this->assertEquals('child1', $node->getNode('childNode2')->getPropertyValue('foo'));
+
+        $this->vm->restore(false, $version);
+        $node = $this->session->getNode($nodePath);
+
+        $this->assertCount(2, $node->getNodes());
+        $this->assertEquals('child1', $node->getNode('childNode1')->getPropertyValue('foo'));
+        $this->assertEquals('child2', $node->getNode('childNode2')->getPropertyValue('foo'));
```

Comment:
ok. the expectations seem correct, this is a problem with jackalope-jackrabbit not noticing that the children also changed. fixed in https://github.com/jackalope/jackalope/pull/294

can you check if things still work out when you have these changes with your code? i wonder if marking nodes as dirty could cause side effects when you restore nodes in memory instead.